File: Searching and Sorting/First and last occurance/firstAndLastOccurance.cpp
#include<iostream>
#include<vector>
using namespace std;

int firstOccurance(vector<int> arr , int el){
     int first = 0;
     int last = arr.size()-1;
     int mid = first + (last - first)/2;
     int firstOccuredIndex;
     while(first <= last){
          if(arr[mid]==el){
               firstOccuredIndex=mid;
               last=mid-1;
          }
          else if(arr[mid]<el){
               first = mid+1;
          }
          else{
               last = mid-1;
          }
          mid = first + (last - first)/2;
     }
     return firstOccuredIndex;
}

int lastOccurance(vector<int> arr , int el){
     int first = 0;
     int last = arr.size()-1;
     int mid = first + (last - first)/2;
     int lastOccuredIndex;
     while(first <= last){
          if(arr[mid]==el){
               lastOccuredIndex=mid;
               first = mid + 1;
          }
          else if(arr[mid]<el){
               first = mid+1;
          }
          else{
               last = mid-1;
          }
          mid = first + (last - first)/2;

     }
     return lastOccuredIndex;
}

int main(){
     vector<int> arr={1,2,3,4,5,5,5,5,6,7,8};
     int el = 5;
     cout<<"first occurance of element at index "<<firstOccurance(arr,el)<<endl;
     cout<<"last occurance of element at index "<<lastOccurance(arr,el)<<endl;
     return 0;
}

File: Searching and Sorting/First and last occurance/firstAndLastOccurance.py
def firstOccurance(arr,el):
     start = 0
     end = len(arr)-1
     mid = start + (end - start)//2
     firstOccuredIndex = -1
     while(start <= end):
          if(arr[mid] == el):
               firstOccuredIndex = mid
               end = mid - 1
          elif(arr[mid]<el):
               start = mid + 1
          else:
               end = mid - 1
          mid = start + (end - start)//2

     return firstOccuredIndex

def lastOccurance(arr,el):
     start = 0
     end = len(arr)-1
     mid = start + (end - start)//2
     lastOccuredIndex = -1
     while(start <= end):
          if(arr[mid] == el):
               lastOccuredIndex = mid
               start = mid + 1
          elif(arr[mid]<el):
               start = mid + 1
          else:
               end = mid - 1
          mid = start + (end - start)//2

     return lastOccuredIndex

def main():
     arr = [1,2,3,4,5,5,5,5,6,7,8]
     el = 5
     print('first occurance is at index ',firstOccurance(arr,el))
     print('last occurance is at index ',lastOccurance(arr,el))

if __name__ == "__main__":
     main()
